kapacitor/kapa_client.go

```diff
@@ -0,0 +1,103 @@
+package kapacitor
+
+import (
+	"sync"
+
+	client "github.com/influxdata/kapacitor/client/v1"
+)
+
+const ListTaskWorkers = 4
+
+// ensure PaginatingKapaClient is a KapaClient
+var _ KapaClient = &PaginatingKapaClient{}
+
+// PaginatingKapaClient is a Kapacitor client that automatically navigates
+// through Kapacitor's pagination to fetch all results
+type PaginatingKapaClient struct {
+	KapaClient
+	FetchRate int // specifies the number of elements to fetch from Kapacitor at a time
+}
+
+// ListTasks lists all available tasks from Kapacitor, navigating pagination as
+// it fetches them
+func (p *PaginatingKapaClient) ListTasks(opts *client.ListTasksOptions) ([]client.Task, error) {
+	// only trigger auto-pagination with Offset=0 and Limit=0
+	if opts.Limit != 0 || opts.Offset != 0 {
+		return p.KapaClient.ListTasks(opts)
+	}
+
+	allTasks := []client.Task{}
+
+	optChan := make(chan client.ListTasksOptions)
+	taskChan := make(chan []client.Task, ListTaskWorkers)
+	done := make(chan struct{})
+
+	var once sync.Once
+
+	doneCloser := func() {
+		close(done)
+	}
+
+	go func() {
+		opts := &client.ListTasksOptions{
+			// copy existing fields
+			TaskOptions: opts.TaskOptions,
+			Pattern:     opts.Pattern,
+			Fields:      opts.Fields,
+
+			// we take control of these two in the loop below
+			Limit:  p.FetchRate,
+			Offset: 0,
+		}
+
+		for {
+			select {
+			case <-done:
```

Comment:
Yep, exactly. You can make `done` a `chan bool`, however the go community considers `chan struct{}`s that you `close` better for pure signals [1]. You also don't want to have a `var done bool` because you would then have to synchronize access to it.

[1] Dave Cheney touched on it [here](https://dave.cheney.net/2013/04/30/curious-channels). This pattern also appears [here](https://blog.golang.org/pipelines) IIRC which is the seminal blog post on Go concurrency patterns.